backend/lucene/src/main/java/org/hibernate/search/backend/lucene/document/impl/LuceneNestedObjectFieldBuilder.java

```diff
@@ -30,7 +30,9 @@ void contribute(MultiTenancyStrategy multiTenancyStrategy, String tenantId, Stri
 		// We own the document content, so we finalize it ourselves.
 		Document document = documentContent.finalizeDocument( multiTenancyStrategy, tenantId, routingKey );
 		document.add( MetadataFields.searchableMetadataField( MetadataFields.typeFieldName(), MetadataFields.TYPE_CHILD_DOCUMENT ) );
-		document.add( MetadataFields.searchableMetadataField( MetadataFields.idFieldName(), rootId ) );
+		// IMPORTANT: Since there is some schema validation now in Lucene the id field of a nested object must match the configuration
+		// of a root document id field. Hence, `searchableRetrievableMetadataField` is used so that the doc_values are set to BINARY as for a root id.
+		document.add( MetadataFields.searchableRetrievableMetadataField( MetadataFields.idFieldName(), rootId ) );
```

Comment:
> there are some schema checks during the writes - when a nested object is written, its ID doesn't match the schema as we were doing this - document.add( MetadataFields.searchableMetadataField( MetadataFields.idFieldName(), rootId ) );. But the ID of a root document has doc-values set to binary. Doing document.add( MetadataFields.searchableRetrievableMetadataField( MetadataFields.idFieldName(), rootId ) ); helps with that.

You mean Lucene 9 has more checks and fails because of this discrepancy?

The thing is, we really don't need docvalues for this field in nested documents, and I wonder what the impact of adding it would be...

Also, I suspect changing this may not fix the problem existing indexes, depending how the thing gets validated. Is it validated globally when Lucene opens the index or only locally for newly indexed documents?

I wonder if we shouldn't just use a different field name for the ID docvalues... We may be able to introduce that without breaking backwards compatibility (try to retrieve the ID from `__HSEARCH_id_docvalue` first, then fall back to `__HSEARCH_id`). Testing this would be annoying, but doable :)

Maybe it would be a good idea to have a look how Elasticsearch does it. If they just enable docvalues for IDs in nested fields, then so be it.

Can you please create a ticket for this in 6.2? Better address this soon.

Reply:
> You mean Lucene 9 has more checks and fails because of this discrepancy?

Yes, that is what it looks like. It happens when a document is indexed. 

I'll take a look into Elasticsearch and then will create a ticket 👍🏻 😃